net: sockets: Add SOCK_RAW support for AF_INET/AF_INET6 sockets

rlubos opened the pull request:
Introduce support for raw IP sockets. 

Fixes https://github.com/zephyrproject-rtos/zephyr/issues/86827

rlubos committed:
net: connection: Register connection type

Register connection type along with family and protocol, so that it's
possible to differentiate between connection listening for raw IP
datagrams and TCP/UDP/other packets.

Signed-off-by: Robert Lubos <robert.lubos@nordicsemi.no>

rlubos committed:
net: sockets: Add SOCK_RAW support for AF_INET/AF_INET6 sockets

Introduce changes in the networking stack which allow to create raw IP
sockets, so that applications can send and receive raw IP datagrams.

Signed-off-by: Robert Lubos <robert.lubos@nordicsemi.no>

rlubos committed:
tests: net: socket: Add tests for AF_INET raw sockets

Add test suite verifying AF_INET/AF_INET6 and SOCK_RAW sockets.

Signed-off-by: Robert Lubos <robert.lubos@nordicsemi.no>

rlubos committed:
doc: release-notes-4-2: Add entry for raw IP sockets

Add bullet point about raw IP sockets support.

Signed-off-by: Robert Lubos <robert.lubos@nordicsemi.no>

jukkar commented:
Just wondering should we have a page in network documentation or doxygen comments in the code that describes how the various options are working (I am referring to the table in the issue of the PR). We mostly follow Linux but there are some deviations so perhaps these should be mentioned somewhere.

rlubos commented:
> Just wondering should we have a page in network documentation or doxygen comments in the code that describes how the various options are working (I am referring to the table in the issue of the PR). We mostly follow Linux but there are some deviations so perhaps these should be mentioned somewhere.

That's a good idea, we have a dedicated documentation page for sockets, perhaps it'd make most sense to add a table in there summarizing what socket types we do support.

jukkar commented on subsys/net/ip/connection.c:
```diff
@@ -635,6 +635,37 @@ static enum net_verdict conn_raw_socket(struct net_pkt *pkt,
+
+	raw_pkt = net_pkt_clone(pkt, CLONE_TIMEOUT);
+	if (!raw_pkt) {
+		return;
```
Cursor will be incorrect now. Perhaps have here goto to the end of the function as the cursor is restored there.

jukkar commented on subsys/net/ip/connection.c:
```diff
@@ -757,6 +797,10 @@ enum net_verdict net_conn_input(struct net_pkt *pkt,
 					continue; /* wrong protocol */
 				}
+			} else if (IS_ENABLED(CONFIG_NET_SOCKETS_INET_RAW) && raw_ip_pkt) {
+				if (conn->proto != 0) {
```
Should we do the check like this
```
if (conn->proto != IPPROTO_IP) {
```
to make it clear what we are checking against (value 0 is a bit vague).

jukkar commented on subsys/net/ip/connection.c:
```diff
@@ -793,6 +837,15 @@ enum net_verdict net_conn_input(struct net_pkt *pkt,
 
 				continue; /* packet was consumed */
 			}
+		} else if (IS_ENABLED(CONFIG_NET_SOCKETS_INET_RAW) && raw_ip_pkt) {
```
Note that few lines above, the code is still checking `IPPROTO_RAW` for packet socket, we need to remove/change those checks from packet socket

rlubos replied:
Yes, do you want me to extend this PR with `AF_PACKET/IPPROTO_RAW` removal?

jukkar replied:
I think we need to remove the `IPPROTO_RAW` from `AF_PACKET`, that was actually the point of the actual issue.

jukkar commented on subsys/net/ip/net_context.c:
```diff
@@ -466,6 +466,9 @@ static int net_context_check(sa_family_t family, enum net_sock_type type,
 		case SOCK_RAW:
+			if (!IS_ENABLED(CONFIG_NET_SOCKETS_INET_RAW)) {
+				NET_DBG("RAW IP sockets disabled.");
+			}
```
Should we return
`return -EPROTONOSUPPORT;`
here?

jukkar commented on subsys/net/ip/net_context.c:
```diff
@@ -2164,6 +2167,57 @@ static int context_setup_udp_packet(struct net_context *context,
+
+		ipv4_hdr = (struct net_ipv4_hdr *)net_pkt_get_data(pkt, &ipv4_access);
+		if (!ipv4_hdr) {
+			return -ENOBUFS;
```
Should we restore the cursor?

jukkar commented on subsys/net/ip/net_context.c:
```diff
@@ -2164,6 +2167,57 @@ static int context_setup_udp_packet(struct net_context *context,
+
+		net_pkt_set_ll_proto_type(pkt, ETH_P_IP);
+	} else {
+		net_pkt_set_ip_hdr_len(pkt, sizeof(struct net_ipv6_hdr));
```
Should we make similar things here as what is being done for IPv4 above?

jukkar commented on subsys/net/lib/sockets/sockets_inet.c:
```diff
@@ -664,6 +680,11 @@ ssize_t zsock_sendmsg_ctx(struct net_context *ctx, const struct msghdr *msg,
 	k_timepoint_t buf_timeout, end;
 	int status;
 
+	if (net_context_get_type(ctx) == SOCK_RAW) {
```
Why do we prevent `sendmsg` for raw sockets?

rlubos replied:
Probably no good reason now that I think of it, I can drop this restriction and add some tests for sendmsg()/recvmsg().

jukkar commented on subsys/net/lib/sockets/sockets_inet.c:
```diff
@@ -1479,6 +1503,11 @@ ssize_t zsock_recvmsg_ctx(struct net_context *ctx, struct msghdr *msg,
 	enum net_sock_type sock_type = net_context_get_type(ctx);
 	size_t i, max_len = 0;
 
+	if (net_context_get_type(ctx) == SOCK_RAW) {
```
Why prevent `recvmsg` for raw sockets?

jukkar reviewed (commented):
Some minor comments.
I am thinking about the tests, how can we verify that we have tested all combinations? Should we have a table in the test describing what we are testing and have a dedicated test function for each "cell" in the table, WDYT?

rlubos commented:
> I am thinking about the tests, how can we verify that we have tested all combinations? Should we have a table in the test describing what we are testing and have a dedicated test function for each "cell" in the table, WDYT?

I can refer to the table from the issue, and extend the test suite a bit. We'll likely duplicate some tests, but that's probably not a big deal. And now I think we should also add tests verifying that all this still works if we enable AF_PACKET support.

rlubos commented:
I have addressed the initial feedback, will work on extending the testsuite even more.

rlubos committed:
doc: net: sockets: Add table summarizing supported socket types

Add a table in sockets documentation summarizing what socket types are
being supported.

Signed-off-by: Robert Lubos <robert.lubos@nordicsemi.no>

rlubos committed:
net: sockets: Remove support for AF_PACKET/IPPROTO_RAW combination

IPPROTO_RAW is not a valid protocol type for AF_PACKET sockets, which
should only use IEEE 802.3 protocol numbers. Therefore remove support
for this type of sockets.

As an alternative, users can use AF_PACKET/SOCK_DGRAM or
AF_INET(6)/SOCK_RAW, depending on the actual use case.

Signed-off-by: Robert Lubos <robert.lubos@nordicsemi.no>

rlubos commented:
In last push:
* Dropped AF_PACKET/IPPROTO_RAW socket support.
* Added table in sockets documentation with the summary of what is supported, let's see how it renders

jukkar reviewed (commented):
The documentation looks great. I think we need to mention in migration doc that the AF_PACKET will not accept IPPROTO_RAW as a protocol any more.

rlubos committed:
doc: migration-guide-4-2: Add entry about dropped IPPROTO_RAW support

Add migration guide entry about dropped IPPROTO_RAW support from
AF_PACKET sockets, with possible alternatives.

Signed-off-by: Robert Lubos <robert.lubos@nordicsemi.no>

rlubos commented:
Ok this should be more or less ready, changes since last week:
* Refined the table in the socket documentation so that it renders better
* Added migration guide entry about removed IPPROTO_RAW support from AF_PACKET
* Added more test cases. I've also added summary in the test suite, based on table created by @jukkar, to make sure that all AF_INET/SOCK_RAW functionalities mentioned there are covered

rlubos commented:
CI fixes

jukkar commented:
@pdgendt please take a look if you have time

jukkar commented:
cc: @go2sh please try if it solves your issues

pdgendt reviewed (dismissed):
Some minor comments

rlubos commented on subsys/net/ip/ipv4.c:
```diff
@@ -383,6 +383,12 @@ enum net_verdict net_ipv4_input(struct net_pkt *pkt, bool is_loopback)
+	ip.ipv4 = hdr;
+
+	if (IS_ENABLED(CONFIG_NET_SOCKETS_INET_RAW)) {
+		(void)net_conn_input(pkt, &ip, hdr->proto, NULL);
```
Technically `net_conn_input()` shouldn't ever fail at that stage, but on the other hands it's probably cleaner to verify the result regardless in case this behavior changes for whatever reason in the future. I've updated this part.

rlubos commented on subsys/net/ip/ipv6.c:
```diff
@@ -762,6 +762,12 @@ enum net_verdict net_ipv6_input(struct net_pkt *pkt, bool is_loopback)
+	ip.ipv6 = hdr;
+
+	if (IS_ENABLED(CONFIG_NET_SOCKETS_INET_RAW)) {
+		(void)net_conn_input(pkt, &ip, current_hdr, NULL);
```
Ditto

rlubos commented on subsys/net/ip/net_context.c:
```diff
@@ -2164,6 +2168,57 @@ static int context_setup_udp_packet(struct net_context *context,
+		net_pkt_set_overwrite(pkt, true);
+
+		ipv4_hdr = (struct net_ipv4_hdr *)net_pkt_get_data(pkt, &ipv4_access);
+		if (!ipv4_hdr) {
```
Fixed
